### ISSUE TITLE
make minimigrator to remove hash_type jinja2 keys

### ISSUE THREAD

**Issue by beckermr** [regro/cf-scripts#847]
These recipes need fixing: https://github.com/search?q=org%3Aconda-forge+hash_type+filename%3Ameta.yaml&type=Code

**Comment by beckermr** [comment]
cc @djhoese

**Comment by djhoese** [comment]
As mentioned in https://github.com/conda-forge/satpy-feedstock/pull/27, it may be good to include removing the `fn` variable usage and the `file_ext` in addition to `hash_type` and `hash_value`.

**Comment by djhoese** [comment]
According to https://github.com/search?q=org%3Aconda-forge+%22set+fn%22+filename%3Ameta.yaml&type=Code that's only 4 repositories with the `fn` usage (but GitHub search has been wrong before).

**Comment by beckermr** [comment]
So see the recipe parser class here: https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/recipe_parser/_parser.py#L308

Using that, it should be fairly straight forward to write the migrator to do the right thing. You may need to delete the extra jinja2 lines at the end. 

If we do this as a mini-migration on top of version, it gets applied first so then the version migrator will also just work.

**Comment by djhoese** [comment]
Looking at the `fn` use cases from the github search I think it is best to ignore this case. Looks like the recipes that use it are doing some python string formatting which might not translate perfectly to pure jinja. Since there are only 6 of them, it is probably fine.

**Comment by djhoese** [comment]
@beckermr Where does this mini-migration actually go? Is it a separate script that I/you put somewhere? I downloaded the `cf-scripts` repository and have it running. I'll make a little script that can do this migration, but I'm not sure where you want it to go.

**Comment by beckermr** [comment]
No sorry. See here: https://github.com/regro/cf-scripts/blob/master/conda_forge_tick/migrators/core.py#L73

This is the base class. I would make a separate file in the migrators package for this. Then import the new migrator up into the top level __init__.py

You will need to add it to the set of migrators attached to the Version migrator in the top of auto_tick.py

You should also add tests in the tests folder using the existing test harnesses we have. See the test_version_migrator.py file for an example.

**Comment by djhoese** [comment]
Ok so:

1. Create `conda_forge_tick/migrators/hash_type_cleanup.py`.
2. In there, subclass from MiniMigrator class and follow other migrators (especially the ones who also use MiniMigrator).
3. Import this class in `conda_forge_tick/migrators/__init__.py`.
4. Add a test to `tests/test_hash_type_cleanup.py`.
5. Make a PR.

Right? If you hate the name of that module, tell me now.

**Comment by beckermr** [comment]
Yup, all seems good to me! Maybe let's call it "ExtraJinja2KeysCleanup"? It should do everything but the fn keys I think. 

When you add the tests, you can stuff any meta.yaml files in the test_yaml folder or w/e it is called.

**Comment by beckermr** [comment]
For the tests, see the pipcheck check tests. The mini-migrator has to piggyback on the version migrator. I would use the satpy feedstock yaml as a test case.

**Comment by djhoese** [comment]
So by doing this as part of the version migrator, this will be run when a package version is updated, right? Put another way, this will not be run with existing package versions?

**Comment by beckermr** [comment]
You are correct. It only gets run on top of other migrations. 

We do this for non-essential tasks in order to keep the traffic down on repos.

**Comment by djhoese** [comment]
Looks like the class would probably have to be added to `auto_tick.py` too? Similar to `CondaForgeYAMLCleanup`?

**Comment by beckermr** [comment]
Yup. See the instructions I left above.

**Comment by djhoese** [comment]
:man_facepalming:

**Comment by beckermr** [comment]
No worries! I just got acquainted with this repo myself and it def takes time!

**Comment by djhoese** [comment]
Anyone have tips for debugging stuff? I'm guessing the `run_test_migration` starts another process that hides all stdout/stderr? Also means a debugger won't work on the test functions. Working around it now by printing to a file.

**Comment by beckermr** [comment]
so use a command like this

`pytest -vvs -x --pdb tests/your_test_file.py`

This will be verbose twice (`-vv`), display stdout/stderr (`-s`), stop at the first failure (`-x') and drop you into pdb ('`-pdb`). For `printf`-debugging, you can use just `-vvs`.

**Comment by djhoese** [comment]
Yeah I was doing `-s` and still nothing from my migrator class. I must be doing something wrong. I'll figure it out. Had to take a break for a couple hours to work on something else, but looking at this now.

**Comment by beckermr** [comment]
seems like the code is not getting called. If you push your changes as you have them now, I may be able to spot the bug and help!

**Comment by djhoese** [comment]
I will try to make a WIP PR in a bit (have to head to a meeting soon). The main thing with trying to use the PyCharm debugger was that I accidentally had it stuck trying to use unittest instead of pytest to load the tests. It was importing the module and then never running the test function.

From the command line and now from my working debugger, the problem is that the version migrator doesn't think the version was updated so no migration was done. Since it detects that it never calls the parent Migrator class's `migrate` method which is the method that calls the piggyback migrators. A lot easier to debug now that my debugger works.

**Comment by djhoese** [comment]
This is the block of code that's causing the issue:

https://github.com/regro/cf-scripts/blob/806f37e0b09070ab41363ace5b51d6dee8656736/conda_forge_tick/migrators/version.py#L439-L450

Right now my test is only updating the version except for the piggy back migrator (which hasn't been run yet) so this always fails.

**Comment by beckermr** [comment]
Please push your PR so I can look. The piggyback migrators get run first, so I don't see why this is causing an issue.

**Comment by beckermr** [comment]
I suspect you are calling the test functions incorrectly

**Comment by djhoese** [comment]
Isn't this actually the whole problem in the first place? The version migrator doesn't properly detect the hash/hash_type so it thinks the package isn't changing. I'll push a PR in a minute (meeting now), but look at the code below the code I linked to above. That code calls `super().migrate(recipe_dir, attrs)` which is where the piggy backs come in to play.

**Comment by beckermr** [comment]
Ahhhhh crap you are right. I had the ordering of the migrators wrong. It calls the piggyback migrations after the version one, not before. This is my mistake and I am very sorry. I think this will have to be a full migrator or we will need to reverse the order of the migrations.

**Comment by djhoese** [comment]
Ok so if this is a full migration. Where should I start? Is my base class a regular `Migrator`? Any existing migration that I should look at as a template?

Since this isn't added to the version stuff, then I probably don't need to touch `auto_tick.py`, right?

**Comment by beckermr** [comment]
I think we should reverse the order of the migrations to put the piggyback ones first.

**Comment by djhoese** [comment]
Ok, sounds good. So I can try this in my PR. You think this can be done in the version migrator's `migrate` method? Right now it looks like it does:

1. Get jinja2 version
2. Update the CondaMetaYAML object's version number (if possible)
3. Dump new YAML to string and compare to old version.
4. Write to `meta.yaml`
5. Call parent class's `migrate` method with mini migrators. The mini migrators all seem to modify the `meta.yaml` file that is on disk. After all piggy back migrators are called, a UID is generated and returned.

Let me try moving things around in the version migrator's `migrate` method and holding on to the UID to return. For the sanity check that some migration besides the version happened (step 3 above) I think I'll have to re-read the `meta.yaml` file from disk and compare it to the old YAML.

**Comment by beckermr** [comment]
We should make the change in the parent class to decouple both. Let me push a PR that does this and fixes another issue.

**Comment by beckermr** [comment]
OK this should fix the issue: https://github.com/regro/cf-scripts/pull/849

you can rebase on master once it is merged

**Comment by djhoese** [comment]
Rebased locally and my migrator is now actually being called! Will try to get something done today for this. Thanks for the quick refactor.

**Comment by djhoese** [comment]
What is the correct way to remove a jinja2 variable from the `CondaMetaYAML` object? If I remove them from `jinja2_vars` they still seem to get written to the `meta.yaml` during the `dump` call. I'll keep looking but if you have any idea what's I'm doing wrong let me know. Basically the version check is still saying that nothing has changed in the meta.yaml even though it really seems like it is. I'll push to my PR later tonight with the changes I've made so far.

**Comment by beckermr** [comment]
So this is not yet supported. I would read the meta.yaml back in and delete the lines. I had support for this but had trouble with it removing lines I needed.

**Comment by djhoese** [comment]
Ok, I'm reworking my code to do this by lines like some of the other migrators. I'm running in to an issue (perhaps with my version of ruamel) where the mappings (dictionaries) in the YAML are showing up as `CommentedMap` objects. This is messing up the version migrator which is doing an `isinstance(src, collections.abc.MutableSequence)` which fails with `CommentedMap`. I haven't changed anything in the version migrator so this must be a YAML reading thing. I'll keep looking tomorrow.

**Comment by beckermr** [comment]
Not sure why this is messing it up. The src section can be a list, so the version migrator looks for lists or for maps. I have tests in the version migrator for both cases.

**Comment by djhoese** [comment]
:man_facepalming: I didn't realize the tests were actually checking the URL and getting the hash. I was using a version that doesn't exist.

(I wonder how many times I'll use that emoji by the time this PR gets finished)

**Comment by beckermr** [comment]
:) 

gotta test it live!
